packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -83,18 +92,28 @@ class CustomDeviceLogReader extends DeviceLogReader {
   ///
   /// Useful when you want to combine the contents of multiple log readers.
   void listenToLinesStream(Stream<String> lines) {
-    _logLinesController.addStream(lines);
+    subscriptions.add(
+      lines.listen(logLinesController.add)
+    );
   }
 
   /// Dispose this log reader, freeing all associated resources and marking
   /// [logLines] as done.
   @override
-  void dispose() {
-    _logLinesController.close();
+  Future<void> dispose() async {
+    final List<Future<void>> futures = <Future<void>>[];
+
+    for (final StreamSubscription<String> subscription in subscriptions) {
+      futures.add(subscription.cancel());
+    }
+
+    futures.add(logLinesController.close());
```

Comment:
Should we not wait for the subscription cancels before closing?

Reply:
Is that a question or a statement? 😄 

Wasn't sure about that, the [docs](https://api.dart.dev/stable/2.14.4/dart-async/StreamSubscription/cancel.html) say the subscription no longer receives events after the call to cancel, however not sure if that means after the returned future completes or after the call returned.

Reply:
Definitely a question :)

Ahh, that's interesting. I had to read it like three times, but it appears to stop receiving events after the initial call. LGTM

Reply:
Okay 👍